sklearn/cluster/_kmeans.py

```diff
@@ -1407,66 +1507,28 @@ def fit(self, X, y=None, sample_weight=None):
             Fitted estimator.
         """
         self._validate_params()
-
-        X = self._validate_data(
+        engine_class = get_engine_class("kmeans", default=KMeansCythonEngine)
+        engine = engine_class(self)
+        X, y, sample_weight = engine.prepare_fit(
             X,
-            accept_sparse="csr",
-            dtype=[np.float64, np.float32],
-            order="C",
-            copy=self.copy_x,
-            accept_large_sparse=False,
+            y=y,
+            sample_weight=sample_weight,
         )
-
         self._check_params_vs_input(X)
 
-        random_state = check_random_state(self.random_state)
-        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
-        self._n_threads = _openmp_effective_n_threads()
-
-        # Validate init array
-        init = self.init
-        init_is_array_like = _is_arraylike_not_scalar(init)
-        if init_is_array_like:
-            init = check_array(init, dtype=X.dtype, copy=True, order="C")
-            self._validate_center_shape(X, init)
-
-        # subtract of mean of x for more accurate distance computations
-        if not sp.issparse(X):
-            X_mean = X.mean(axis=0)
-            # The copy was already done above
-            X -= X_mean
-
-            if init_is_array_like:
-                init -= X_mean
-
-        # precompute squared norms of data points
-        x_squared_norms = row_norms(X, squared=True)
-
-        if self._algorithm == "elkan":
-            kmeans_single = _kmeans_single_elkan
-        else:
-            kmeans_single = _kmeans_single_lloyd
-            self._check_mkl_vcomp(X, X.shape[0])
-
         best_inertia, best_labels = None, None
 
         for i in range(self._n_init):
             # Initialize centers
-            centers_init = self._init_centroids(
-                X, x_squared_norms=x_squared_norms, init=init, random_state=random_state
-            )
+            centers_init = engine.init_centroids(X)
             if self.verbose:
                 print("Initialization complete")
 
             # run a k-means once
-            labels, inertia, centers, n_iter_ = kmeans_single(
+            labels, inertia, centers, n_iter_ = engine.kmeans_single(
                 X,
                 sample_weight,
                 centers_init,
-                max_iter=self.max_iter,
-                verbose=self.verbose,
-                tol=self._tol,
-                n_threads=self._n_threads,
             )
```

Comment:
Note that currently we delegate the full Lloyd loop to the engine for the sake of simplicity for this first iteration.

However my plan would rather be to only delegate one Lloyd iteration at a time instead and move the loop and convergence check back into the estimator class (probably in a new private method).

Moving this loop back into the estimators class will be required to properly integrate @jeremiedbb's work on callbacks for instance: #22000.

Reply:
While looking at the `KMeans` in cuml and sklearn I noticed that my test data set that took ~2.5s for sklearn took only 25ms in cuml. I assume with numba-dpex the speed up will be similar. So the main point is "something that takes long now might be super fast in a plugin".

On the one hand `max_iter` is typically(?) only O(100) so the overhead of making a few hundred additional function calls is not that great. On the other hand, what data would the callbacks need to make decisions (beyond just progressbar updates, which you could argue you don't need for fast stuff) and how much does it cost to transfer that from device to device?

My question is: do you/someone know what happens to the performance of alternative implementations of kmeans when they have to be implemented as "one iteration at a time"? I don't know enough about how kmeans on something like a GPU works, but I hope the answer is "nothing happens, this is totally doable" because that would make life easy.

Reply:
In general it will be implemented "one interation at a time" (exemple with the plugin we're working on: https://github.com/soda-inria/sklearn-numba-dpex/blob/main/sklearn_numba_dpex/kmeans/drivers.py#L342, or likewise with the current cython implementation in sklearn: https://github.com/scikit-learn/scikit-learn/blob/17df37aee/sklearn/cluster/_kmeans.py#L647) and the loop on `max_iter` is in python, it is completely negligible.

Reply:
I expect that that for most algorithms in scikit-learn, one fit iteration would last 10x or more the overhead of a GPU dispatch (assuming we leave the data and temporary datastructures on device memory between consecutive iterations).

Reply:
What is the plan regarding this change? I think it would make sense to already go for the "one call per iteration" (what you propose) now instead of waiting. If that works for you I'll send a PR.

Reply:
It will also help sharpen the API for when data is passed in to an engine. My proposal would to settle on:

* `prepare_fit(X, y, sample_weight)` will be called every time when `KMeans.fit()` is called. The engine should keep hold of the data as it won't be provided again
* the single iteration method becomes `kmeans_single(current_centroids)`. `X` and friends isn't passed again.

In particular when we call `kmeans_single` once per iteration we need a way to provide the data to the engine once, so that it can convert it or decide that it would like to raise `NotImplemented`, etc. If `X` and friends are passed to `kmeans_single` on each iteration, the engine should in principle be prepared to convert it at each iteration because there is no promise that it hasn't change since the last time `kmeans_single` was called.

Something that would be harder to do in that case is re-use things from the Cython engine. For example currently I use:
```python
def init_centroids(self, X):
    return cp.asarray(super().init_centroids(X))
```
because there is not much benefit in re-implementing the initial centroid selection. In a super strict world where the above proposal is adopted, then `init_centroids` should also not be passed `X` again, but it should use what was previously provided in `prepare_fit()`.

The point being, I'd support switching to the per-iteration API so we can work these things out with code (instead of our heads).

Reply:
I think we agree on the implementation strategy here so you're welcome to send a PR  :+1: and let's iterate on the details from there if needed, I think @ogrisel @jjerphan and I can be available for review.

Regarding:

> because there is not much benefit in re-implementing the initial centroid selection.

Actually I can think of two reasons:

- initial centroid selection includes k-means++ which might benefit from optimizations

- it's not likely that 'super().init_centroids(X)` will work if `X` is not a numpy or scipy array, and it could be a good thing for the `fit` to accept other type of inputs when used with other engines, e.g `cupy` arrays for a a cuda-based engine, to save data loading to and from the targeted device. So, in 'super().init_centroids(X)` might need to be rewritten for cases where `X` is a different object.